File: 10. TowerOfHanoi.cpp
/*
Task: Implementation and Time Analysis of Tower of Hanoi (Recursive)
*/
#include <iostream>
#include <ctime>
using namespace std;

void towerOfHanoi(int n, char from, char to, char aux)
{
    if (n == 0)
        return;
    towerOfHanoi(n - 1, from, aux, to);
    cout << "Move disk " << n << " from " << from << " to " << to << endl;
    towerOfHanoi(n - 1, aux, to, from);
}
int main()
{
    int n = 3;
    clock_t start = clock();
    towerOfHanoi(n, 'A', 'C', 'B');
    clock_t end = clock();
    cout << "Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    return 0;
}
/*
Time Complexity: O(2^n)
Space Complexity: O(n)
*/

File: 10. TowerOfHanoi.py
"""
Task: Implementation and Time Analysis of Tower of Hanoi (Recursive)
"""
def tower_of_hanoi(n, from_peg, to_peg, aux_peg):
    if n == 0:
        return
    tower_of_hanoi(n-1, from_peg, aux_peg, to_peg)
    print(f"Move disk {n} from {from_peg} to {to_peg}")
    tower_of_hanoi(n-1, aux_peg, to_peg, from_peg)
if __name__ == "__main__":
    import time
    n = 3
    start = time.time()
    tower_of_hanoi(n, 'A', 'C', 'B')
    end = time.time()
    print(f"Time: {end-start:.6f}s")
"""
Time Complexity: O(2^n)
Space Complexity: O(n)
"""
